h2o-algos/src/main/java/hex/glrm/GLRM.java

```diff
@@ -1131,13 +1189,37 @@ private Chunk chk_xnew(Chunk[] chks, int c) {
 
           // Gradient wrt x_i is matrix product \grad L_{i,j}(x_i * Y_j, A_{i,j}) * Y_j'
           double[] weight = _lossFunc[j].mlgrad(xy, (int) a[j]);
-          double[][] ysub = _yt.getCatBlock(j);
-          for (int k = 0; k < _ncolX; k++) {
-            for (int c = 0; c < weight.length; c++)
-              grad[k] += cweight * weight[c] * ysub[k][c];
+ //         double[][] ysub = _yt.getCatBlock(j);
+          if (_yt._transposed) {
+            for (int k = 0; k < _ncolX; k++) {
+              for (int c = 0; c < weight.length; c++) {
+                //             grad[k] += cweight * weight[c] * ysub[k][c];
+                int cidx = _yt.getCatCidx(j, c);
+                double archtypevalues = 0;
+                archtypevalues = _yt._archetypes[cidx][k];
+                tgrad[k] += cweight * weight[c] * archtypevalues;
+              }
+            }
```

Comment:
Ok, so since you've split this into 2 cases, why not take it 1 step further:
change the order of loops (i.e. outer loop in `c`, inner in `k`), then `cidx` will need to be computed only once in the outer loop, and you can even cache the values of `_yt._archetypes[cidx]` and `cweight * weight[c]`